cclib/method/ddec.py

```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
+            grid = ((x, y, z) for x in range(xshape) for y in range(yshape) for z in range(zshape))
+            for xindex, yindex, zindex in grid:
+                # leftterm is the first spherical average term in equation 66.
+                # <rho^cond_A(r_A) / sqrt(rho^cond(r))>
+                self._leftterm[atomi][xindex][yindex][zindex] = self._cond_density[atomi][
+                    self.closest_r_index[atomi][xindex][yindex][zindex]
+                ]
+            self._leftterm[atomi] = self._leftterm[atomi] / rho_cond_sqrt
+            for radiusi in range(len(self.tau[atomi])):
+                grid_filter = self.closest_r_index[atomi] == radiusi
+                num_grid_filter = numpy.count_nonzero(grid_filter)
+                if num_grid_filter < 1:
+                    self.tau[atomi][radiusi] = 0.0
+                else:
+                    leftaverage = numpy.sum(grid_filter * self._leftterm[atomi]) / num_grid_filter
+                    rightaverage = numpy.sum(grid_filter * rho_cond_sqrt) / num_grid_filter
+                    if leftaverage < 1e-20:
+                        self.tau[atomi][radiusi] = 0.0
+                    else:
+                        self.tau[atomi][radiusi] = numpy.divide(
+                            leftaverage,
+                            rightaverage,
+                            out=numpy.zeros_like(leftaverage),
+                            where=rightaverage != 0.0,
+                        )
+            # Make tau monotonic decreasing
+            self.tau[atomi] = numpy.maximum.accumulate(self.tau[atomi][::-1])[::-1]
+
+    def _psiai(self, ya, bigpsiAI, atomi):
+        # Re-evaluate cond_density
+        cond_density = ya + bigpsiAI * numpy.sqrt(ya)
+
+        # Monotonic Decrease Condition
+        cond_density = numpy.minimum.accumulate(cond_density)
+
+        # Re-evaluate psi_AI
+        psiAI = (
+            self._integrate_from_radial([cond_density], [atomi])
+            - self.data.atomnos[atomi]
+            + self.refcharges[-1][atomi]
+        )
+
+        return psiAI, cond_density
+
+    def _ya(self, proatom_density, atomi):
+        # Function that calculates Y_a^avg
+        # See Eq. 40-41 in doi: 10.1039/c6ra04656h
+        rho_ref = self._rho_ref
+        # Y_a^avg -- Equation 40 in doi: 10.1039/c6ra04656h
+        ya = numpy.zeros_like(proatom_density[atomi], dtype=float)
+        weights = self.chgdensity.data / rho_ref
+        summed_elec = 0
+        for radiusi in range(len(ya)):
+            grid_filter = self.closest_r_index[atomi] == radiusi
+            num_grid_filter = numpy.count_nonzero(grid_filter)
+            if num_grid_filter < 1:
+                ya[radiusi] = 0.0
+            else:
+                spherical_avg = numpy.sum(grid_filter * weights) / num_grid_filter
+                ya[radiusi] = proatom_density[atomi][radiusi] * spherical_avg
+
+        ya = numpy.maximum.accumulate(ya[::-1])[::-1]
```

Comment:
Can you explain why you need to reverse, accumulate, then reverse again?

Reply:
This was in the `module_reshaping_functions.f08:77-79`. I've added a comment about this. Thanks.